File: 백준/Silver/1012. 유기농 배추/유기농 배추.py
from collections import deque
answer = []

for _ in range(int(input())):
    M, N, K = map(int, input().split())
    maps = [[0] * (N + 1) for i in range(M + 1)]
    visited = [[False] * (N + 1) for i in range(M + 1)]
    result = []

    for _ in range(K):
        X, Y = map(int, input().split())
        maps[X][Y] = 1

    def bfs(x, y):
        queue = deque()
        queue.append((x, y))

        dx = [-1, 1, 0, 0]
        dy = [0, 0, -1, 1]

        cnt = 0

        while queue:
            x, y = queue.popleft()

            for i in  range(4):
                nx = x + dx[i]
                ny = y + dy[i]

                if 0 <= nx < M and 0 <= ny < N:
                    if maps[nx][ny] == 1 and visited[nx][ny] == False:
                        queue.append((nx, ny))
                        visited[nx][ny] = True
                        cnt += 1
        return cnt


    for i in range(M):
        for j in range(N):
            if maps[i][j] == 1 and visited[i][j] == False:
                visited[i][j] = True
                result.append(bfs(i, j))

    answer.append(len(result))
    
for i in answer:
    print(i)

File: 백준/Silver/1012. 유기농 배추/유기농 배추.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;
import java.util.Queue;
import java.util.LinkedList;

public class Main {
    static int T, N, M, K;
    static int result;
    static int[][] maps;
    static boolean[][] visited;
    static Queue<int[]> q;
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};


    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        StringBuilder sb = new StringBuilder();

        T = Integer.parseInt(st.nextToken());

        for (int i = 0; i < T; i++) {
            st = new StringTokenizer(br.readLine());
            M = Integer.parseInt(st.nextToken());
            N = Integer.parseInt(st.nextToken());
            K = Integer.parseInt(st.nextToken());

            maps = new int[M][N];
            visited = new boolean[M][N];


            for (int j = 0; j < K; j++) {
                st = new StringTokenizer(br.readLine());
                int x = Integer.parseInt(st.nextToken());
                int y = Integer.parseInt(st.nextToken());

                maps[x][y] = 1;
            }


            result = 0;

            for (int nm = 0; nm < M; nm++) {
                for (int nn = 0; nn < N; nn++) {
                    if (maps[nm][nn] == 1 && !visited[nm][nn]) {
                        bfs(nm, nn);
                        result++;
                    }
                }
            }

            sb.append(result).append("\n");
        }

        System.out.println(sb);

    }


    public static void bfs(int startX, int startY) {
        q = new LinkedList<>();
        q.add(new int[]{startX, startY});
        visited[startX][startY] = true;

        while (!q.isEmpty()) {
            int[] cuurent = q.poll();
            int x = cuurent[0];
            int y = cuurent[1];

            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];

                if (nx >= 0 && ny >= 0 && nx < M && ny < N && !visited[nx][ny] && maps[nx][ny] == 1) {
                    q.add(new int[]{nx, ny});
                    visited[nx][ny] = true;
                }
            }
        }
    }
}
